### PR TITLE
Normalize expressions.

### LINKED ISSUE #1943: sql: normalize expressions
Add a `NormalizeExpr` pass which normalizes an expression tree. Examples of normalizations to perform:

* Replace `Row` with `Tuple` which are synonymous allowing expression evaluation to only worry about `Tuple`.
* Replace `ParenExpr` with its contents. The parentheses are only necessary for proper parsing, not for evaluation.
* Replace branches of the tree which evaluate to a constant with those constants. For example, `a+1+1` should be transformed into `a+2`.

There are likely many more such transformations. One medium-term challenge is to figure out how to write these transformations, but as a start, ad-hoc code will get the ball rolling.

### PR DESCRIPTION
**Pull request by petermattis**
Added NormalizeExpr which performs various bits of expression
normalization. The most important of these normalizations are the
evaluation of constant branches of the expression and the rewriting of
expressions so that "variables" are isolated on the left sides of
comparisons.

Fixes #1943.

### CODE REVIEW

File: sql/parser/walk.go [hunk 1]
```diff
@@ -23,14 +23,18 @@ import "fmt"
 // WalkExpr. The returned Expr replaces the pointer to the visited expression
```
Comment (tamird) [inline sql/parser/walk.go]: this comment needs an update

File: sql/parser/walk.go [hunk 2]
```diff
@@ -23,14 +23,18 @@ import "fmt"
...
 // in the parent node and can be used for rewriting expressions.
 type Visitor interface {
-	Visit(expr Expr) Expr
+	Visit(expr Expr, pre bool) (Visitor, Expr)
 }
 
 // WalkExpr traverses the nodes in an expression. It starts by calling
-// v.Visit(expr). It then recursively traverses the children nodes of the
-// expression returned by v.Visit().
+// v.Visit(expr, true). If the visitor returned by v.Pre(expr) is not nil it
```
Comment (tamird) [inline sql/parser/walk.go]: is .Pre a thing?

File: sql/parser/walk.go [hunk 3]
```diff
@@ -23,14 +23,18 @@ import "fmt"
 // WalkExpr. The returned Expr replaces the pointer to the visited expression
 // in the parent node and can be used for rewriting expressions.
 type Visitor interface {
-	Visit(expr Expr) Expr
+	Visit(expr Expr, pre bool) (Visitor, Expr)
```
Comment (tamird) [inline sql/parser/walk.go]: so `pre` is just an alias for "not base case"?

File: sql/parser/normalize.go [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+			v.isConst = false
+			return nil, expr
+		}
+	}
+	return v, expr
+}
+
+// IsConst returns true if the expression contains only constant values
+// (i.e. it does not contain a DReference).
+func IsConst(expr Expr) bool {
```
Comment (tamird) [inline sql/parser/normalize.go]: should this be public?

**Comment by tamird** [comment]
LGTM modulo nits, but I'll admit I didn't pour over every line of this.

File: sql/parser/walk.go (hunk 1 above)
Comment (petermattis) [inline sql/parser/walk.go]: Done.

File: sql/parser/walk.go (hunk 3 above)
Comment (petermattis) [inline sql/parser/walk.go]: Short for `preorder`, but that probably isn't the right terminology. Let me think if there is a better term.

File: sql/parser/walk.go (hunk 2 above)
Comment (petermattis) [inline sql/parser/walk.go]: Well, it was, and then I added the `pre` argument to `Visit`. Fixed the comment.

File: sql/parser/normalize.go (hunk 4 above)
Comment (petermattis) [inline sql/parser/normalize.go]: It was used by index selection during the development of the PR, but that has since changed. I've unexported it.

**Pull request merged**